File: 0042-trapping-rain-water/0042-trapping-rain-water.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        int n = height.size();
        int total = 0;
        int lMax = height[0];
        int rMax = height[n-1];

        int l = 0;
        int r = height.size() - 1;

        while(l < r) {
            if(height[l] <= height[r]) {
                if(lMax > height[l]) {
                    total += lMax - height[l];
                }else{
                    lMax = height[l];
                }
                l = l+1;
            }
            else{
                if(rMax > height[r]) {
                    total += rMax - height[r];
                }else{
                    rMax = height[r];
                }
                r = r - 1;
            }
        }
        return total;
    }
};

File: 0042-trapping-rain-water/0042-trapping-rain-water.java
class Solution {
    public int trap(int[] height) {
        int n = height.length;

        int[] prefMax = new int[n];
        int maxi = Integer.MIN_VALUE;

        for(int i=0; i<n; i++) {
            maxi = Math.max(maxi, height[i]);
            prefMax[i] = maxi;
        }

        int[] suffMax = new int[n];
        maxi = Integer.MIN_VALUE;
        for(int i=n-1; i>=0; i--) {
            maxi = Math.max(maxi, height[i]);
            suffMax[i] = maxi;
        }

        int total = 0;
        for(int i=0; i<n; i++) {
            total += (Math.min(prefMax[i], suffMax[i]) - height[i]);
        }
        return total;
    }
}